crates/futures/src/queue.rs

```diff
@@ -0,0 +1,90 @@
+use js_sys::Promise;
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+use std::rc::Rc;
+use wasm_bindgen::prelude::*;
+
+pub(crate) trait Task {
+    fn run(self);
+}
+
+struct QueueState<A> {
+    // This is used to ensure that it's only scheduled once
+    is_spinning: Cell<bool>,
+
+    // This is a queue of Tasks which will be run in order
+    tasks: RefCell<VecDeque<A>>,
+}
+
+impl<A> QueueState<A>
+where
+    A: Task,
+{
+    fn run_all(&self) {
+        loop {
```

Comment:
Could a comment here be added that this greedily performs all work, but we may want to one day in the future have it capped at a fixed number of iterations to prevent an infinite loop in wasm from never returning to the event loop?

Reply:
That's how the microtask event queue works in general, you can definitely create infinite loops with JS Promises which never return to the event loop:

```js
function loop() {
    return Promise.resolve().then(() => {
        return loop();
    });
}

loop();
```

Do we want to diverge from Promise semantics?

(I also believe that's how non-wasm Executors work, they can also get into infinite loops with their event loop)

Reply:
Er sorry I don't mean that we'll actively break anything, only try to make executions a bit more fair. If you write an infinite loop you'll write an infinite loop, but a typically thing with executors like this is to yield back to the system every so often to let other items progress if that's the case.

Reply:
But doing any sort of yielding would diverge from Promise semantics. I'm not sure that's a good idea, I'd rather the user insert a manual `wait(0)` or similar if they want to yield.

Reply:
Er so just to be clear, you understand that I'm not asking for code to change, I'm just asking for a comment, right? This is something we're going to want to get to eventually because it's basically what every futures executor does, so I'd rather not debate its merits now but rather we can deal with that if the need for an implementation actually arises. I'd just like some documentation indicating that this is possibly a known problem in case someone runs into it in the future.